core/src/jsMain/kotlin/Peripheral.kt

```diff
@@ -52,9 +50,12 @@ public class JsPeripheral internal constructor(
     private val _state = MutableStateFlow<State?>(null)
     public override val state: Flow<State> = _state.filterNotNull()
 
-    private var platformServices: List<PlatformService>? = null
+    private var _platformServices: List<PlatformService>? = null
+    private val platformServices: List<PlatformService>
+        get() = checkNotNull(_platformServices) { "Services have not been discovered for $this" }
+
     public override val services: List<DiscoveredService>?
-        get() = platformServices?.map { it.toDiscoveredService() }
+        get() = _platformServices?.map { it.toDiscoveredService() }
```

Comment:
Yes, following the guideline put forth in [Error-handling style and exceptions](https://github.com/Kotlin/KEEP/blob/master/proposals/stdlib/result.md#error-handling-style-and-exceptions):

> In general, if some API requires its callers to handle failures locally (immediately around or next to the invocation), then it should use nullable types, when these failures do not carry additional business meaning, or domain-specific data types to represent its successful results and failures with any additional business-related data that is needed to process these failures.

So by making `services` nullable, allows caller to decide what (if any) exception to throw if services have not been discovered. By not throwing, saves us the expensive operation of creation an exception as well.

This is described in the `Peripheral` interface so that library consumers are made aware of the conditions in which `services` would be `null`.

https://github.com/JuulLabs/kable/blob/9a4e186eddbcde7881c9ed2b4b4b9a620e5a40c9/core/src/commonMain/kotlin/Peripheral.kt#L63-L64

Although, I acknowledge that it does feel much like an error condition to try and retrieve the services before services have been discovered, so I'm open to change it if makes more sense to throw.

Reply:
Got it, thanks for clarifying. I should have checked the header docs in the interface definition for the expected behavior.